### PR TITLE
Added "auto select best class" feature for qx.data.marshal.Json

### PR DESCRIPTION
**Pull request by schnepe2**
Attempt to fix issue #9193

### CODE REVIEW

**Comment by level420** [comment]
@schnepe2 Thank you for the PR.

It seems that one of your tests fail constantly on Firefox and Chrome:
```
Chrome 53.0.2785 (Linux 0.0.0) qx.test.data.store Json:testOwnClassWith FAILED
	Not the given class used!: Expected 'qx.test.AB' but found 'qx.data.model.a~b'!
		qx.core.AssertionError.constructor:43:38
		qx.core.Assert.__fail:71:19
		qx.core.Assert.assertEquals:177:33
		qx.type.BaseString.assertEquals:87:22
		anonymous()
		qx.dev.unit.TestResult.run:237:36
		qx.dev.unit.TestCase.resume:100:35
		anonymous()
		qx.event.dispatch.Direct.dispatchEvent:134:34
		wrappedFunction()
		qx.event.Manager.prototype.dispatchEvent()
		qx.event.Registration.fireNonBubblingEvent()
		qx.core.Object.prototype.fireDataEvent()
		qx.data.store.Json.prototype._onSuccess()
		...
```

**Comment by schnepe2** [comment]
That "old" test assumes a name for the class that might not be correct.
Strange that this didn't happen on my machine. These tests are head-less, right? I'll see what I have to do to make them pass again.

**Comment by schnepe2** [comment]
The test that failed **now** (qx.test.ui.basic Label:testApplyWebFont) is nothing I have done (I think)

**Comment by level420** [comment]
@schnepe2 Yes the failing test ```testApplyWebFont``` is failing because of timing issues with travis. See https://github.com/qooxdoo/qooxdoo/issues/9202

But there is another failure here:
```
Running lint...
  - Warning: qx.data.marshal.Json (118, 11): Declared but unused variable or parameter: 'clazz'
Lint warnings are treated as errors! Please fix!
```

**Comment by schnepe2** [comment]
About all the tests that failed **now**: I have no idea what happens there.
And unfortunately I have spend more time on this than my boss would allow me to do, so either someone from the core team looks over this or we have to wait until I can steal some more time.

**Comment by schnepe2** [comment]
F***k ! I hate the complexity of git! It distracts me from the most obvious things!
I'll try a fix of the fix...

**Comment by level420** [comment]
The PR is passing the tests. Start voting/approval.

File: framework/source/class/qx/data/marshal/Json.js
Comment (johnspackman) [inline framework/source/class/qx/data/marshal/Json.js]: Please can you use a different name for the variable - the underscore does not help describe it's purpose, eg `bubbleClassName` ?

File: framework/source/class/qx/data/marshal/Json.js
Comment (johnspackman) [inline framework/source/class/qx/data/marshal/Json.js]: also please rename the underscores here

**Review by johnspackman — CHANGES_REQUESTED** [review]

**Comment by schnepe2** [comment]
@johnspackman Done (not tested of cause!).
Although I think those kind of qooxdoo-internal naming conventions could also have been easily fixed in trunk by you :wink:

**Comment by oetiker** [comment]
@schnepe2 we do not fix things in master ... changes get into master via PR ... and sure, someone might have done a cleanup PR after merging yours, but then again ... since we have no cleanup crew on staff ...

**Review by johnspackman — APPROVED** [review]

**Review by level420 — APPROVED** [review]

**Comment by level420** [comment]
@schnepe2 If I (qooxdoo member) want to change something in master, I have to go through the same procedure you've gone. Create a PR ... Wait for travis runs to be happy ... Start voting ... Wait for comments or change requests ... modify ... sync the PR ...  

The only thing I can do what you can't is merging. But NOT my own PRs.

**Comment by schnepe2** [comment]
The corner case of a very simple model (having only one property (like `{a:"a"}`) ) wasn't working
(no glue character to distinguish what kind of class we have)
I therefore changed the "decoration" of the hash by adding a heart prefix ("♥") to model-classes which support the bubble event.

**Review by oetiker — APPROVED** [review]

**Comment by oetiker** [comment]
@schnepe2 have you seen how many fixes I made to my #9194 PR until it was merged? We do this for two reasons:

a) to ensure that qx master is always clean (to the best of our knowledge), so whenever you pull master you can be sure you get 'the good stuff'

b) often we have to merge the patches for master also with the current stable branch (aka 5.0.x) if every change is one PR, then this is much easier.

**Review by cajus — APPROVED** [review]

**Comment by level420** [comment]
4 approvals, more than 2 business days
merging
Thank you @schnepe2

**Pull request merged**
